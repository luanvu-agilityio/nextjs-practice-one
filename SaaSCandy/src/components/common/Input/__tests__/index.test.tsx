import { render, screen, fireEvent } from '@testing-library/react';
import { Input } from '../index';
import { Mail } from 'lucide-react';

describe('Input - Snapshots', () => {
  it('should match snapshot with default props', () => {
    const { container } = render(
      <Input label='Email' placeholder='Enter email' />
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with error state', () => {
    const { container } = render(
      <Input
        id='test-input-error'
        label='Email'
        placeholder='Enter email'
        errorMessage='Invalid email'
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with password toggle', () => {
    const { container } = render(
      <Input
        label='Password'
        type='password'
        showPasswordToggle
        placeholder='Enter password'
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with left element', () => {
    const { container } = render(
      <Input
        label='Search'
        placeholder='Search...'
        leftElement={<Mail className='h-4 w-4' />}
      />
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Input - Interactive', () => {
  it('should handle text input changes', () => {
    const handleChange = jest.fn();
    render(
      <Input label='Email' placeholder='Enter email' onChange={handleChange} />
    );

    const input = screen.getByLabelText('Email');
    fireEvent.change(input, { target: { value: 'test@example.com' } });

    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(input).toHaveValue('test@example.com');
  });

  it('should toggle password visibility', () => {
    render(
      <Input
        label='Password'
        type='password'
        showPasswordToggle
        placeholder='Enter password'
      />
    );

    const input = screen.getByLabelText('Password');
    const toggleButton = screen.getByLabelText('Show password');

    expect(input).toHaveAttribute('type', 'password');

    fireEvent.click(toggleButton);
    expect(input).toHaveAttribute('type', 'text');
    expect(screen.getByLabelText('Hide password')).toBeInTheDocument();

    fireEvent.click(screen.getByLabelText('Hide password'));
    expect(input).toHaveAttribute('type', 'password');
  });

  it('should display error message when provided', () => {
    const errorMessage = 'This field is required';
    render(
      <Input
        label='Email'
        placeholder='Enter email'
        errorMessage={errorMessage}
      />
    );

    expect(screen.getByText(errorMessage)).toBeInTheDocument();
    const input = screen.getByLabelText('Email');
    expect(input).toHaveAttribute('aria-invalid', 'true');
  });

  it('should be disabled when disabled prop is true', () => {
    render(<Input label='Email' placeholder='Enter email' disabled />);

    const input = screen.getByLabelText('Email');
    expect(input).toBeDisabled();
  });

  it('should hide label visually when hideLabel is true', () => {
    render(<Input label='Email' placeholder='Enter email' hideLabel />);

    const label = screen.getByText('Email');
    expect(label).toHaveClass('sr-only');

    const input = screen.getByPlaceholderText('Enter email');
    expect(input).toHaveAttribute('aria-label', 'Email');
  });

  it('should use custom id when provided', () => {
    const customId = 'custom-input-id';
    render(<Input id={customId} label='Email' placeholder='Enter email' />);

    const input = screen.getByLabelText('Email');
    expect(input).toHaveAttribute('id', customId);
  });
});
