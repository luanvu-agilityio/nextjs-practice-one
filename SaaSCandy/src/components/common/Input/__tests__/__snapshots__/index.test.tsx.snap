// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Input - Snapshots > should match snapshot with default props 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="input-«r0»"
    >
      Email
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-gray-background"
    >
      <input
        aria-invalid="false"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="input-«r0»"
        placeholder="Enter email"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`Input - Snapshots > should match snapshot with error state 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="test-input-error"
    >
      Email
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-destructive ring-destructive/20"
    >
      <input
        aria-invalid="true"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="test-input-error"
        placeholder="Enter email"
        type="text"
      />
    </div>
    <p
      class="text-md text-destructive-background"
      role="paragraph"
    >
      Invalid email
    </p>
  </div>
</div>
`;

exports[`Input - Snapshots > should match snapshot with left element 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="input-«r3»"
    >
      Search
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-gray-background"
    >
      <span
        class="flex mr-2 text-gray-foreground"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-mail h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m22 7-8.991 5.727a2 2 0 0 1-2.009 0L2 7"
          />
          <rect
            height="16"
            rx="2"
            width="20"
            x="2"
            y="4"
          />
        </svg>
      </span>
      <input
        aria-invalid="false"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="input-«r3»"
        placeholder="Search..."
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`Input - Snapshots > should match snapshot with password toggle 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="input-«r2»"
    >
      Password
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-gray-background"
    >
      <input
        aria-invalid="false"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="input-«r2»"
        placeholder="Enter password"
        type="password"
      />
      <span
        class="flex ml-2 text-gray-foreground"
      >
        <button
          aria-label="Show password"
          class="p-1 text-gray-foreground hover:text-gray-700 transition-colors"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-eye h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </button>
      </span>
    </div>
  </div>
</div>
`;

exports[`Input - Snapshots should match snapshot with default props 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="input-«r0»"
    >
      Email
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-gray-background"
    >
      <input
        aria-invalid="false"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="input-«r0»"
        placeholder="Enter email"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`Input - Snapshots should match snapshot with error state 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="test-input-error"
    >
      Email
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-destructive ring-destructive/20"
    >
      <input
        aria-invalid="true"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="test-input-error"
        placeholder="Enter email"
        type="text"
      />
    </div>
    <p
      class="text-md text-destructive-background"
      role="paragraph"
    >
      Invalid email
    </p>
  </div>
</div>
`;

exports[`Input - Snapshots should match snapshot with left element 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="input-«r3»"
    >
      Search
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-gray-background"
    >
      <span
        class="flex mr-2 text-gray-foreground"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-mail h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m22 7-8.991 5.727a2 2 0 0 1-2.009 0L2 7"
          />
          <rect
            height="16"
            rx="2"
            width="20"
            x="2"
            y="4"
          />
        </svg>
      </span>
      <input
        aria-invalid="false"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="input-«r3»"
        placeholder="Search..."
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`Input - Snapshots should match snapshot with password toggle 1`] = `
<div>
  <div
    class="flex flex-col gap-2 w-full"
  >
    <label
      class="text-sm font-medium text-placeholder"
      for="input-«r2»"
    >
      Password
    </label>
    <div
      class="flex items-center h-11 w-full min-w-0 rounded-lg border text-md transition-colors bg-transparent px-4 py-2.5 focus-within:border-white border-gray-background"
    >
      <input
        aria-invalid="false"
        class="w-full bg-transparent outline-none placeholder:text-placeholder disabled:pointer-events-none disabled:cursor-not-allowed"
        id="input-«r2»"
        placeholder="Enter password"
        type="password"
      />
      <span
        class="flex ml-2 text-gray-foreground"
      >
        <button
          aria-label="Show password"
          class="p-1 text-gray-foreground hover:text-gray-700 transition-colors"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-eye h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
        </button>
      </span>
    </div>
  </div>
</div>
`;
